### PR TITLE
Sync.poll doesn't retry if the request fails for some reason :…

### LINKED ISSUE #1164: Sync.poll doesn't retry if the request fails for some reason.
**Describe the bug**
In the Demo Application, if user presses `Sync` from the Slider Navigation and the request fails for some reason (internet not there or connection failed), engine doesn't retry syncing again until the next periodic sync happens.

**Component**
Demo App

**To Reproduce**

**Expected behavior**
The engine should retry as per the `PeriodicSyncConfiguration.RetryConfiguration`.
**Screenshots**

**Smartphone (please complete the following information):**
 - Device: [e.g. Pixel4a emulator]
 - Android version: [e.g. Settings -> About phone -> Android version]
 - Build number: [e.g. Settings -> About phone -> Build number]

**Additional context**

**Would you like to work on the issue?**
Please state if this issue should be assigned to you or who you think could help to solve this issue.

### PR DESCRIPTION
**Pull request by PallaviGanorkar**
… bug fixed

Fixes #1164

**Description**
Sync.poll doesn't retry if the request fails for some reason : bug fixed
Solution -  Along with Periodic sync, BackoffCriteria is added to retry sync request.

**Type**
Choose one: Bug fix 

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

**Comment by aditya-07** [comment]
@PallaviGanorkar Please use **Description** section above to describe your solution.

File: engine/src/main/java/com/google/android/fhir/sync/SyncJobImpl.kt [hunk 1]
```diff
@@ -60,16 +61,28 @@ class SyncJobImpl(private val context: Context) : SyncJob {
...
+          periodicSyncConfiguration.retryConfiguration.backoffCriteria.backoffPolicy,
+          periodicSyncConfiguration.retryConfiguration.backoffCriteria.backoffDelay,
+          periodicSyncConfiguration.retryConfiguration.backoffCriteria.timeUnit
+        )
+        .setInputData(
+          Data.Builder()
+            .putInt(MAX_RETRIES_ALLOWED, periodicSyncConfiguration.retryConfiguration.maxRetries)
+            .build()
+        )
+    }
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/sync/SyncJobImpl.kt:77]: `RetryConfiguration.backoffCriteria` is `non-null` type.
```suggestion
    periodicSyncConfiguration.retryConfiguration?.let {
      periodicWorkRequest
        .setBackoffCriteria(
       it.backoffCriteria.backoffPolicy,
        it.backoffCriteria.backoffDelay,
        it.backoffCriteria.timeUnit
        )
        .setInputData(
          Data.Builder()
            .putInt(MAX_RETRIES_ALLOWED, it.maxRetries)
            .build()
        )
    }
```

**Review by aditya-07 — CHANGES_REQUESTED** [review]
Please add appropriate tests to check that the retry mechanism is working.

**Comment by PallaviGanorkar** [comment]
> @PallaviGanorkar Please use **Description** section above to describe your solution.



> @PallaviGanorkar Please use **Description** section above to describe your solution.

Updated Description

**Comment by Tarun-Bhardwaj** [comment]
@PallaviGanorkar , are the changes ready so that @aditya-07  can review?

**Comment by PallaviGanorkar** [comment]
@aditya-07  Please review the code changes, As discussed, reusing existing method for periodic sync with retry mechanism

**Comment by PallaviGanorkar** [comment]
> @PallaviGanorkar Please use **Description** section above to describe your solution.

done

**Comment by PallaviGanorkar** [comment]
> Please add appropriate tests to check that the retry mechanism is working.

@aditya-07  As discussed, Code changes are tested Manually, those are working fine.
Test case will be pushed as separate PR under issue https://github.com/google/android-fhir/issues/1252

File: engine/src/main/java/com/google/android/fhir/sync/Sync.kt [hunk 2]
```diff
@@ -64,40 +62,43 @@ object Sync {
    * Starts a one time sync based on [FhirSyncWorker]. In case of a failure, [RetryConfiguration]
    * will guide the retry mechanism. Caller can set [retryConfiguration] to [null] to stop retry.
    */
-  inline fun <reified W : FhirSyncWorker> oneTimeSync(
+  inline fun <W : FhirSyncWorker> oneTimeSync(
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/sync/Sync.kt:65]: This function signature shouldn't change. Instead you should internally call `createOneTimeWorkRequest(retryConfiguration, W::class.java)`

File: engine/src/main/java/com/google/android/fhir/sync/Sync.kt [hunk 3]
```diff
@@ -64,40 +62,43 @@ object Sync {
...
     context: Context,
+    clazz: Class<W>,
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/sync/Sync.kt:67]: remove it.

File: demo/src/main/java/com/google/android/fhir/demo/FhirApplication.kt [hunk 4]
```diff
@@ -43,7 +43,7 @@ class FhirApplication : Application() {
         ServerConfiguration("https://hapi.fhir.org/baseR4/")
       )
     )
-    Sync.oneTimeSync<FhirPeriodicSyncWorker>(this)
+    Sync.oneTimeSync(this, FhirPeriodicSyncWorker::class.java)
```
Comment (aditya-07) [inline demo/src/main/java/com/google/android/fhir/demo/FhirApplication.kt:46]:
```suggestion
    Sync.oneTimeSync<FhirPeriodicSyncWorker>(this)
```
Revert this change as the  Sync.oneTimeSync signature shouldn't change.

File: engine/src/main/java/com/google/android/fhir/sync/Sync.kt [hunk 5]
```diff
@@ -64,40 +62,43 @@ object Sync {
...
   /**
    * Starts a periodic sync based on [FhirSyncWorker]. It takes [PeriodicSyncConfiguration] to
    * determine the sync frequency and [RetryConfiguration] to guide the retry mechanism. Caller can
    * set [retryConfiguration] to [null] to stop retry.
    */
-  inline fun <reified W : FhirSyncWorker> periodicSync(
+  inline fun <W : FhirSyncWorker> periodicSync(
     context: Context,
     periodicSyncConfiguration: PeriodicSyncConfiguration,
+    clazz: Class<W>
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/sync/Sync.kt:85]:
```suggestion
  inline fun < reified W : FhirSyncWorker> periodicSync(
    context: Context,
    periodicSyncConfiguration: PeriodicSyncConfiguration)
```
Just like oneTimeSync, this api shouldn't change as well.

File: engine/src/main/java/com/google/android/fhir/sync/Sync.kt [hunk 6]
```diff
@@ -64,40 +62,43 @@ object Sync {
...
   ) {
 
     WorkManager.getInstance(context)
       .enqueueUniquePeriodicWork(
         SyncWorkType.DOWNLOAD.workerName,
         ExistingPeriodicWorkPolicy.KEEP,
-        createPeriodicWorkRequest<W>(periodicSyncConfiguration)
+        createPeriodicWorkRequest(periodicSyncConfiguration, clazz)
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/sync/Sync.kt:92]:
```suggestion
        createPeriodicWorkRequest(periodicSyncConfiguration, W::class.java)
```

File: engine/src/main/java/com/google/android/fhir/sync/Sync.kt [hunk 7]
```diff
@@ -64,40 +62,43 @@ object Sync {
...
     context: Context,
+    clazz: Class<W>,
     retryConfiguration: RetryConfiguration? = defaultRetryConfiguration
   ) {
     WorkManager.getInstance(context)
       .enqueueUniqueWork(
         SyncWorkType.DOWNLOAD.workerName,
         ExistingWorkPolicy.KEEP,
-        createOneTimeWorkRequest<W>(retryConfiguration)
+        createOneTimeWorkRequest(retryConfiguration, clazz)
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/sync/Sync.kt:74]:
```suggestion
        createOneTimeWorkRequest(retryConfiguration, W::class.java)
```

File: engine/src/main/java/com/google/android/fhir/sync/Sync.kt [hunk 8]
```diff
@@ -64,40 +62,43 @@ object Sync {
...
 
   @PublishedApi
-  internal inline fun <reified W : FhirSyncWorker> createOneTimeWorkRequest(
-    retryConfiguration: RetryConfiguration?
+  internal inline fun <W : FhirSyncWorker> createOneTimeWorkRequest(
+    retryConfiguration: RetryConfiguration?,
+    clazz: Class<W>
   ): OneTimeWorkRequest {
-    val oneTimeWorkRequest = OneTimeWorkRequestBuilder<W>()
+    val oneTimeWorkRequest = OneTimeWorkRequest.Builder(clazz)
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/sync/Sync.kt:101]: While you are at it, please rename the variable.
```suggestion
    val oneTimeWorkRequestBuilder = OneTimeWorkRequest.Builder(clazz)
```

File: engine/src/test/java/com/google/android/fhir/sync/SyncTest.kt [hunk 9]
```diff
@@ -42,8 +42,9 @@ class SyncTest {
   @Test
   fun createOneTimeWorkRequestWithRetryConfiguration_shouldHave3MaxTries() {
     val workRequest =
-      Sync.createOneTimeWorkRequest<PassingPeriodicSyncWorker>(
-        RetryConfiguration(BackoffCriteria(BackoffPolicy.LINEAR, 30, TimeUnit.SECONDS), 3)
+      Sync.createOneTimeWorkRequest(
+        RetryConfiguration(BackoffCriteria(BackoffPolicy.LINEAR, 30, TimeUnit.SECONDS), 3),
+        PassingPeriodicSyncWorker::class.java
```
Comment (aditya-07) [inline engine/src/test/java/com/google/android/fhir/sync/SyncTest.kt:47]: Should be reverted.

File: engine/src/test/java/com/google/android/fhir/sync/SyncTest.kt [hunk 10]
```diff
@@ -52,7 +53,7 @@ class SyncTest {
 
   @Test
   fun createOneTimeWorkRequest_withoutRetryConfiguration_shouldHaveZeroMaxTries() {
-    val workRequest = Sync.createOneTimeWorkRequest<PassingPeriodicSyncWorker>(null)
+    val workRequest = Sync.createOneTimeWorkRequest(null, PassingPeriodicSyncWorker::class.java)
```
Comment (aditya-07) [inline engine/src/test/java/com/google/android/fhir/sync/SyncTest.kt:56]: Should be reverted.

File: engine/src/test/java/com/google/android/fhir/sync/SyncTest.kt [hunk 11]
```diff
@@ -61,12 +62,13 @@ class SyncTest {
   @Test
   fun createPeriodicWorkRequest_withRetryConfiguration_shouldHave3MaxTries() {
     val workRequest =
-      Sync.createPeriodicWorkRequest<PassingPeriodicSyncWorker>(
+      Sync.createPeriodicWorkRequest(
```
Comment (aditya-07) [inline engine/src/test/java/com/google/android/fhir/sync/SyncTest.kt:65]: Should be reverted.

File: engine/src/test/java/com/google/android/fhir/sync/SyncTest.kt [hunk 12]
```diff
@@ -77,11 +79,12 @@ class SyncTest {
...
   fun createPeriodicWorkRequest_withoutRetryConfiguration_shouldHaveZeroMaxRetries() {
     val workRequest =
-      Sync.createPeriodicWorkRequest<PassingPeriodicSyncWorker>(
+      Sync.createPeriodicWorkRequest(
         PeriodicSyncConfiguration(
           repeat = RepeatInterval(20, TimeUnit.MINUTES),
           retryConfiguration = null
-        )
+        ),
+        PassingPeriodicSyncWorker::class.java
```
Comment (aditya-07) [inline engine/src/test/java/com/google/android/fhir/sync/SyncTest.kt:87]: Should be reverted.

**Review by aditya-07 — CHANGES_REQUESTED** [review]
Left some comments. We shouldn't change our public `oneTimeSync` and `periodicSync` apis. They should still take `reified` type.

**Comment by PallaviGanorkar** [comment]
> As discussed, Code changes are tested Manually, those are working fine.
> Test case will be pushed as separate PR under issue #1252

As discussed, Code changes are tested Manually, those are working fine.
Test case will be added as separate PR under issue https://github.com/google/android-fhir/issues/1252

File: engine/src/main/java/com/google/android/fhir/sync/Sync.kt (hunk 6 above)
Comment (PallaviGanorkar) [reply engine/src/main/java/com/google/android/fhir/sync/Sync.kt]: This changes request not valid for internal method

File: engine/src/main/java/com/google/android/fhir/sync/Sync.kt (hunk 7 above)
Comment (PallaviGanorkar) [reply engine/src/main/java/com/google/android/fhir/sync/Sync.kt]: This changes request not valid for internal method

File: engine/src/test/java/com/google/android/fhir/sync/SyncTest.kt (hunk 9 above)
Comment (PallaviGanorkar) [reply engine/src/test/java/com/google/android/fhir/sync/SyncTest.kt:48]: This changes request not valid for internal method

File: engine/src/test/java/com/google/android/fhir/sync/SyncTest.kt (hunk 10 above)
Comment (PallaviGanorkar) [reply engine/src/test/java/com/google/android/fhir/sync/SyncTest.kt:57]: This changes request not valid for internal method

File: engine/src/test/java/com/google/android/fhir/sync/SyncTest.kt (hunk 11 above)
Comment (PallaviGanorkar) [reply engine/src/test/java/com/google/android/fhir/sync/SyncTest.kt:66]: This changes request not valid for internal method

File: engine/src/test/java/com/google/android/fhir/sync/SyncTest.kt (hunk 12 above)
Comment (PallaviGanorkar) [reply engine/src/test/java/com/google/android/fhir/sync/SyncTest.kt:88]: This changes request not valid for internal method

File: engine/src/main/java/com/google/android/fhir/sync/Sync.kt [hunk 13]
```diff
@@ -72,7 +70,7 @@ object Sync {
       .enqueueUniqueWork(
         SyncWorkType.DOWNLOAD.workerName,
         ExistingWorkPolicy.KEEP,
-        createOneTimeWorkRequest<W>(retryConfiguration)
+        createOneTimeWorkRequest<W>(retryConfiguration, W::class.java)
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/sync/Sync.kt:73]:
```suggestion
        createOneTimeWorkRequest(retryConfiguration, W::class.java)
```

**Review by aditya-07 — APPROVED** [review]
Just one small cleanup, otherwise LGTM!

**Comment by PallaviGanorkar** [comment]
> Just one small cleanup, otherwise LGTM!

Done

**Pull request merged**